### PR TITLE
better API version handling - part 1

### PR DESCRIPTION
**Pull request by fearthecowboy**

### CODE REVIEW

File: modelerfour/flattener/flattener.ts [hunk 1]
```diff
@@ -193,6 +193,13 @@ export class Flattener {
 
       for (const group of this.codeModel.operationGroups) {
         for (const operation of group.operations) {
+          // when there are multiple requests in an operation
+          // and the
```
Comment (daviwil) [inline modelerfour/flattener/flattener.ts:197]: Missing the rest of the sentence here I think

File: modelerfour/modeler/modelerfour.ts [hunk 2]
```diff
@@ -84,6 +127,62 @@ export class ModelerFour {
...
+    this.preprocessOperations();
+  }
+
+  preprocessOperations() {
+    // preprocess to get all http operations flattend out into a nice neat collection
+    for (const { key: path, value: pathItem } of this.resolveDictionary(this.input.paths)) {
+      for (const httpMethod of [HttpMethod.Delete, HttpMethod.Get, HttpMethod.Head, HttpMethod.Options, HttpMethod.Patch, HttpMethod.Post, HttpMethod.Put, HttpMethod.Trace]) {
+        const httpOperation = pathItem[httpMethod];
+        if (httpOperation) {
+          this.inputOperations.push({
```
Comment (daviwil) [inline modelerfour/modeler/modelerfour.ts:139]: What does it mean to be an "InputOperation" if all of the operations get added to the list?

File: modelerfour/modeler/modelerfour.ts [hunk 3]
```diff
@@ -84,6 +127,62 @@ export class ModelerFour {
...
+
+  initApiVersionMode(apiVersionParameter: 'choice' | 'constant' | undefined, useModelNamespace: boolean | undefined) {
+    if (this.profileFilter.length > 0) {
+      // must be profile mode.
+      return 'profile';
+    }
+
+    // see how many api versions there are for all the operations
+    const allApiVersions = values(this.inputOperations).selectMany(each => {
+      const apiv = <Array<string>>this.interpret.xmsMetaFallback(each.operation, each.pathItem, 'apiVersions');
```
Comment (daviwil) [inline modelerfour/modeler/modelerfour.ts:158]: Should the key at the end be `apiVersions` or `apiVersion`?

File: modelerfour/modeler/modelerfour.ts [hunk 4]
```diff
@@ -84,6 +127,62 @@ export class ModelerFour {
...
+    }).distinct().toArray();
+    switch (allApiVersions.length) {
+      case 0:
+        this.useModelNamespace = false;
+        return 'none';
+
+      case 1:
+        this.apiVersionParameter = apiVersionParameter || 'constant';
+        this.useModelNamespace = useModelNamespace || false;
+        return 'client';
```
Comment (daviwil) [inline modelerfour/modeler/modelerfour.ts:170]: What if > 1?

File: modelerfour/modeler/modelerfour.ts [hunk 5]
```diff
@@ -549,7 +658,8 @@ export class ModelerFour {
 
     if (!isMoreThanObject && !hasProperties) {
       // it's an empty object?
-      this.session.warning(`Schema '${name}' is an empty object without properties or modifiers.`, ['Modeler', 'EmptyObject'], aSchema);
+      // this.session.warning(`Schema '${name}' is an empty object without properties or modifiers.`, ['Modeler', 'EmptyObject'], aSchema);
+      return this.anySchema;
```
Comment (daviwil) [inline modelerfour/modeler/modelerfour.ts:662]: Returning the schema seems to be a distinct change in behavior, was it missing before or is there a reason to add it now?

File: modelerfour/test/scenarios/a-playground/flattened.yaml [hunk 6]
```diff
@@ -1675,7 +1675,7 @@ schemas: !<!Schemas>
           default:
             name: MyDerivedType
             description: ''
-            namespace: Api20160229
```
Comment (daviwil) [inline modelerfour/test/scenarios/a-playground/flattened.yaml:1678]: So should we expect to see all namespaces go away until multiple API versions are detected?

Comment (fearthecowboy) [reply modelerfour/test/scenarios/a-playground/flattened.yaml:1678]: They can pass `--use-model-namespace:false` to turn them back on. It default to the logic in #173 to figure out what to do.

File: modelerfour/modeler/modelerfour.ts (hunk 5 above)
Comment (fearthecowboy) [reply modelerfour/modeler/modelerfour.ts:662]: We've re-declared `{type:object}`  to mean "any"  *sigh*

File: modelerfour/modeler/modelerfour.ts (hunk 4 above)
Comment (fearthecowboy) [reply modelerfour/modeler/modelerfour.ts:170]: it passes out of the switch.

File: modelerfour/modeler/modelerfour.ts (hunk 3 above)
Comment (fearthecowboy) [reply modelerfour/modeler/modelerfour.ts:158]: It's `apiVersions` (in the xmsMetadata, we collect the apiversions for the node and toss them in an array)

File: modelerfour/modeler/modelerfour.ts (hunk 2 above)
Comment (fearthecowboy) [reply modelerfour/modeler/modelerfour.ts:139]: it distinguishes it from where I used the work `operation` and `httpOperation` already in the same method.

File: modelerfour/flattener/flattener.ts (hunk 1 above)
Comment (fearthecowboy) [reply modelerfour/flattener/flattener.ts:197]: This is why we don't code past 8pm.

Comment (fearthecowboy) [reply modelerfour/flattener/flattener.ts:197]: correcting

**Pull request merged**